pyirf/interpolation/moment_morph_interpolator.py

```diff
@@ -0,0 +1,442 @@
+        # cs = np.broadcast_to(cs, mus.shape)
+        mu_strich = np.sum(cs * mus, axis=0)
+        sig_strich = np.sum(cs * sigs, axis=0)
+
+        # Compute slope and offset as in eq. (14) and (15) in [1]
+        aij = sigs / sig_strich
+        bij = mus - mu_strich * aij
+
+        # Transformation as in eq. (13) in [1]
+        mids = np.broadcast_to(self.bin_mids, self.bin_contents.shape)
+        transf_mids = aij[..., np.newaxis] * mids + bij[..., np.newaxis]
+
+        # Compute the morphed historgram according to eq. (18) in [1]. The function "lookup" "resamples"
+        # the histogram at the transformed bin-mids by using the templates historgam value at the transformed
+        # bin-mid as new value for a whole transformed bin. Second adaption needed to extend
+        # the moment morph procedure to histograms, adaptes the behaviour of eq. (16)
+
+        transf_hist = self._lookup(transf_mids)
+
+        f_new = np.sum(
+            np.expand_dims(cs, -1) * transf_hist * np.expand_dims(aij, -1), axis=0
+        )
+
+        # Reset interpolation resolts for those templates with partially zero entries from above to 0
+        f_new[zero_templates] = np.zeros(len(self.bin_mids))
+
+        # If used normally, all values are stricktly positive, if used for extrapolation, some values can
+        # become slightly negative as the mean/std estimation is not exact.
+        f_new[f_new < 0] = 0
+
+        # Re-Normalize, needed, as the estimation of the std used above is not exact but the result is scaled with
+        # the estimated std
+        norm = np.expand_dims(np.sum(f_new, axis=-1), -1)
+
+        return np.divide(
+            f_new, norm, out=np.zeros_like(f_new), where=norm != 0
+        ).reshape(1, *self.bin_contents.shape[1:])
+
+
+class Base2DTriangularMomentMorphInterpolator(BaseMomentMorphInterpolator):
+    def __init__(self, grid_points, bin_edges, bin_contents):
+        """
+        Base class implementing the MomentMorphInterpolation in a 2D triangle.
+
+        Parameters
+        ----------
+        grid_points: np.ndarray, shape=(3, 2)
+            Grid points at which interpolation templates exist
+        bin_edges: np.ndarray, shape=(M+1)
+            Edges of the data binning
+        bin_content: np.ndarray, shape=(3, ..., M)
+            Content of each bin in bin_edges for
+            each point in grid_points. First dimesion has to correspond to number
+            of grid_points, last dimension has to correspond to number of bins for
+            the quantity that should be interpolated (e.g. the Migra axis for EDisp)
+
+        Note
+        ----
+            Also calls BaseMomentMorphInterpolator's __init__.
+        """
+        if grid_points.shape != (3, 2):
+            raise ValueError("This base class can only interpolate in a triangle.")
+
+        super().__init__(grid_points, bin_edges, bin_contents)
+
+    def _target_in_grid(self, target_point):
```

Comment:
This is bad: _target_in_grid now "lies" to circumvent a check that is not needed for something that can extrapolate.

Reply:
If you want to keep the base class structure, I think it would be better to keep the `_target_in_grid` method untouched, since it is well-defined.

Instead, change the check where necessary to something like:

```
if not self.can_extrapolate and not self._target_in_grid(target_point):
    raise ValueError("Cannot extrapolate")
```